src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,178 @@
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+locale chain_sync =
+  fixes initial_items\<^sub>c :: "'i list"
+  fixes initial_items\<^sub>s :: "'i list"
+  fixes candidate_points :: "'i list \<Rightarrow> 'p list"
+  fixes best_intersection_point :: "'i list \<Rightarrow> 'p list \<Rightarrow> 'p option"
+  fixes point_of :: "'i \<Rightarrow> 'p"
+
+subsection \<open>Parties\<close>
+
+datatype party =
+  Client |
+  Server
+
+subsection \<open>State Machine\<close>
+
+datatype state =
+  Idle |
+  Intersect |
+  CanAwait
+
+datatype ('i, 'p) message =
+  is_next_request: RequestNext |
+  is_roll_forward: RollForward \<open>'i\<close> |
+  is_roll_backward: RollBackward \<open>'p\<close> |
+  is_await_reply: AwaitReply |
+  is_find_intersect: FindIntersect \<open>'p list\<close> |
+  is_intersect_found: IntersectFound \<open>'p\<close> |
+  is_intersect_not_found: IntersectNotFound
+
+fun agent_in_state' where
+  "agent_in_state' Idle = Client" |
+  "agent_in_state' _ = Server"
+
+inductive can_finish_in_state' where
+  "can_finish_in_state' Idle"
+
+declare can_finish_in_state'.simps [simp]
+
+primrec next_state' where
+  "next_state' Idle m = (partial_case m of
+    FindIntersect _ \<Rightarrow> Intersect |
+    RequestNext \<Rightarrow> CanAwait
+  )" |
+  "next_state' Intersect m = (partial_case m of
+    IntersectFound _ \<Rightarrow> Idle |
+    IntersectNotFound \<Rightarrow> Idle
+  )" |
+  "next_state' CanAwait m = (partial_case m of
+    RollForward _ \<Rightarrow> Idle |
+    RollBackward _ \<Rightarrow> Idle |
+    AwaitReply \<Rightarrow> Idle \<comment> \<open>only for this initial implementation\<close>
+  )"
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_items_update: ItemsUpdate
+
+definition roll_back_to :: "'i list \<Rightarrow> 'p \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> 'i list" where
+  [simp]: "roll_back_to \<I> p \<psi> = (THE \<I>\<^sub>1. \<exists>\<I>\<^sub>2. \<I> = \<I>\<^sub>1 @ \<I>\<^sub>2 \<and> \<psi> (last \<I>\<^sub>1) = p)"
+
+corec client_program where
+  "client_program ph \<kappa> \<psi> \<I> = (
+    case ph of
+      IntersectionFinding \<Rightarrow>
+        \<up> Cont (FindIntersect (\<kappa> \<I>));
+        \<down> M; (partial_case M of
+          Cont IntersectNotFound \<Rightarrow>
+            \<up> Done;
+            \<bottom> |
+          Cont (IntersectFound _) \<Rightarrow>
+            client_program ItemsUpdate \<kappa> \<psi> \<I>
```

Comment:
The intersection point is thrown away by the client because the server will nevertheless send it to the client in a `RollBackward` message as the response to the first `RequestNext` sent by the client. This is actually what _The Shelley Networking Protocol_ states:

> Whenever the server replies with MsgIntersectFound the client can expect the next
update (i.e. a replay to MsgRequestNext) to be MsgRollBackward to the specified pointintersect (which makes handling state updates on the client side easier)

What do you think?

Reply:
I think that you’re absolutely right. :slightly_smiling_face: No change necessary here apparently. :+1: